codegen/src/RenderPure.hs

```diff
@@ -0,0 +1,175 @@
+isRename (BindRename _ _) = True
+isRename _ = False
+
+removeBinding :: Binding -> (String, D.Declaration) -> Bool
+removeBinding (BindRemove n) (hsName, _) = n == hsName
+removeBinding _ _ = False
+
+removeBinding' :: [Binding] -> (String, D.Declaration) -> Bool
+removeBinding' bindings decl = any (\b -> removeBinding b decl) bindings
+
+removeFilter :: [Binding] -> [(String, D.Declaration)] -> [(String, D.Declaration)]
+removeFilter bindings fns = filter (\v -> not (removeBinding' bindings' v)) fns
+  where
+    bindings' = filter isRemove bindings
+
+renameBinding :: Binding -> (String, D.Declaration) -> Maybe (String, D.Declaration)
+renameBinding (BindRename n new_name) (hsName, decl) =
+  if n == hsName then Just (new_name,decl) else Nothing
+renameBinding _ _ = Nothing
+
+renameBinding' :: [Binding] -> (String, D.Declaration) -> (String, D.Declaration)
+renameBinding' bindings decl@(_,d) =
+  case foldl (\i b -> let v = (renameBinding b decl) in if isJust v then v else i) Nothing bindings of
+    Just v -> v
+    Nothing -> (D.name d,d)
+
+renameFilter ::  [Binding] -> [(String, D.Declaration)] -> [(String, D.Declaration)]
+renameFilter bindings fns = map (renameBinding' bindings') fns
+  where
+    bindings' = filter isRename bindings
+
+nativeFunctionsFilter :: [D.Declaration] -> [Binding] -> [(String, D.Declaration)]
+nativeFunctionsFilter fns bindings =
+  filter (\(_,a) ->
+            D.mode a == D.Native &&
+            "namespace" `elem` (D.method_of a) &&
+            D.is_factory_method a == Nothing &&
+            not (isPrefixOf "_" (D.name a)) &&
+            not (isSuffixOf "_" (D.name a)) &&
+            not (isSuffixOf "_out" (D.name a)) &&
+            all (/= P.Ptr P.GeneratorType) (map D.dynamic_type' (D.arguments a))
+         ) $
+  renameFilter bindings $
+  removeFilter bindings $
+  map (\f -> (getSignatures (toFunction f),f)) fns
+
+notUniqList :: [String] -> [String]
+notUniqList lst = notUniq (sort lst) []
+  where
+    notUniq [] a = a
+    notUniq (x:y:xs) ys = if x == y then notUniq xs (y:ys) else (notUniq (y:xs) ys)
+    notUniq _ b = b
+
+decodeAndCodeGen :: String -> String -> String -> IO ()
+decodeAndCodeGen basedir yamlSpecFileName bindingsFileName = do
+  funcs <- Y.decodeFileEither yamlSpecFileName :: IO (Either ParseException [D.Declaration])
+  bindings <- Y.decodeFileEither bindingsFileName :: IO (Either ParseException [Binding])
+  case (funcs,bindings) of
+    (Left err', _) -> print err'
+    (Right _  , Left err') -> print err'
+    (Right fns, Right bnd) -> do
+      createDirectoryIfMissing True (basedir <> "/Torch/Functions/")
+      let l = nativeFunctionsFilter fns bnd
+
+      case notUniqList (map fst l) of
+        [] -> do 
+          T.writeFile (basedir <> "/Torch/Functions/Native.hs") $
+            template "Torch.Functions.Native" $
+            renderFunctions l
+        xs -> do
+          putStrLn "---Duplicated functions are as follows. ----"
+          forM_ xs $ \x -> do
```

Comment:
When codegen finds duplicated definitions of functions like sum_t and sum_tt, it stops generating codes and outputs the names of functions.